R/lazyframe-frame.R

```diff
@@ -358,12 +755,1924 @@ lazyframe__slice <- function(offset, length = NULL) {
+#' ser <- lf$serialize()
+#' pl$deserialize_lf(ser)
+pl__deserialize_lf <- function(source) {
+  wrap({
+    deserialize_lf(source)
+  })
+}
+
+#' Explode the DataFrame to long format by exploding the given columns
+#'
+#' @param ... <[`dynamic-dots`][rlang::dyn-dots]> Column names, expressions, or
+#' a selector defining them. The underlying columns being exploded must be of
+#' the `List` or `Array` data type.
+#'
+#' @inherit as_polars_lf return
+#' @examples
+#' lf <- pl$LazyFrame(
+#'   letters = c("a", "a", "b", "c"),
+#'   numbers = list(1, c(2, 3), c(4, 5), c(6, 7, 8))
+#' )
+#'
+#' lf$explode("numbers")$collect()
+lazyframe__explode <- function(...) {
+  wrap({
+    check_dots_unnamed()
+    by <- parse_into_list_of_expressions(...)
+    self$`_ldf`$explode(by)
+  })
+}
+
+#' Clone a LazyFrame
+#'
+#' This makes a very cheap deep copy/clone of an existing
+#' [`LazyFrame`][lazyframe__class]. Rarely useful as `LazyFrame`s are nearly 100%
+#' immutable. Any modification of a `LazyFrame` should lead to a clone anyways,
+#' but this can be useful when dealing with attributes (see examples).
+#'
+#'
+#' @inherit as_polars_lf return
+#' @examples
+#' df1 <- as_polars_lf(iris)
+#'
+#' # Make a function to take a LazyFrame, add an attribute, and return a LazyFrame
+#' give_attr <- function(data) {
+#'   attr(data, "created_on") <- "2024-01-29"
+#'   data
+#' }
+#' df2 <- give_attr(df1)
+#'
+#' # Problem: the original LazyFrame also gets the attribute while it shouldn't!
+#' attributes(df1)
+#'
+#' # Use $clone() inside the function to avoid that
+#' give_attr <- function(data) {
+#'   data <- data$clone()
+#'   attr(data, "created_on") <- "2024-01-29"
+#'   data
+#' }
+#' df1 <- as_polars_lf(iris)
+#' df2 <- give_attr(df1)
+#'
+#' # now, the original LazyFrame doesn't get this attribute
+#' attributes(df1)
+lazyframe__clone <- function() {
+  self$`_ldf`$clone()
+}
+
+
+#' Decompose struct columns into separate columns for each of their fields
+#'
+#' The new columns will be inserted into the LazyFrame at the location of the
+#' struct column.
+#'
+#' @param ... <[`dynamic-dots`][rlang::dyn-dots]> Name of the struct column(s)
+#' that should be unnested.
+#'
+#' @inherit as_polars_lf return
+#' @examples
+#' lf <- pl$LazyFrame(
+#'   a = 1:5,
+#'   b = c("one", "two", "three", "four", "five"),
+#'   c = 6:10
+#' )$
+#'   select(
+#'   pl$struct("b"),
+#'   pl$struct(c("a", "c"))$alias("a_and_c")
+#' )
+#' lf$collect()
+#'
+#' lf$unnest("a_and_c")$collect()
+#' lf$unnest(pl$col("a_and_c"))$collect()
+lazyframe__unnest <- function(...) {
+  wrap({
+    check_dots_unnamed()
+    columns <- parse_into_list_of_expressions(...)
+    self$`_ldf`$unnest(columns)
+  })
+}
+
+#' Add an external context to the computation graph
+#'
+#' This allows expressions to also access columns from DataFrames or LazyFrames
+#' that are not part of this one.
+#'
+#' @param other Data/LazyFrame to have access to. This can be a list of DataFrames
+#' and LazyFrames.
```

Comment:
Is this really true?

Reply:
`with_context()` was actually deprecated in 1.0, I'll remove it